### PR TITLE
Support `None` values in array, map and row types

### LINKED ISSUE #269: TypeError: float() argument must be a string or a number, not 'NoneType' when using experimental_python_types=True
### Expected behavior

None values are mapped to `None` in `_array_map_func`

E.g

```
    def _array_map_func(self, column):
        element_mapping_func = self._col_func(column['arguments'][0]['value'])
        return lambda values: [element_mapping_func(value) if value is not None else None for value in values]
```

### Actual behavior

None in arrays cause `TypeError: float() argument must be a string or a number, not 'NoneType'`

### Steps To Reproduce

query a real array column containing a mix of `None` and real values

### Log output

```
  File "/usr/local/lib/python3.9/site-packages/trino/dbapi.py", line 435, in execute
    result = self._query.execute()
  File "/usr/local/lib/python3.9/site-packages/trino/client.py", line 758, in execute
    self._result.rows += self.fetch()
  File "/usr/local/lib/python3.9/site-packages/trino/client.py", line 782, in fetch
    return self._row_mapper.map(status.rows)
  File "/usr/local/lib/python3.9/site-packages/trino/client.py", line 982, in map
    return [self._map_row(row) for row in rows]
  File "/usr/local/lib/python3.9/site-packages/trino/client.py", line 982, in <listcomp>
    return [self._map_row(row) for row in rows]
  File "/usr/local/lib/python3.9/site-packages/trino/client.py", line 985, in _map_row
    return [self._map_value(value, self.columns[idx]) for idx, value in enumerate(row)]
  File "/usr/local/lib/python3.9/site-packages/trino/client.py", line 985, in <listcomp>
    return [self._map_value(value, self.columns[idx]) for idx, value in enumerate(row)]
  File "/usr/local/lib/python3.9/site-packages/trino/client.py", line 992, in _map_value
    return col_mapping_func(value)
  File "/usr/local/lib/python3.9/site-packages/trino/client.py", line 892, in <lambda>
    return lambda values: [element_mapping_func(value) for value in values]
  File "/usr/local/lib/python3.9/site-packages/trino/client.py", line 892, in <listcomp>
    return lambda values: [element_mapping_func(value) for value in values]
  File "/usr/local/lib/python3.9/site-packages/trino/client.py", line 907, in <lambda>
    else float(val)
TypeError: float() argument must be a string or a number, not 'NoneType'

```

### Operating System

Bullseye

### Trino Python client version

trino-0.318.0

### Trino Server version

367

### Python version

3.9

### Are you willing to submit PR?

- [X] Yes I am willing to submit a PR!

### PR DESCRIPTION
**Pull request by mdesmet**
## Description

Fixes #269

Also refactors mappers from lambda functions to separate classes for better readability.

## Release notes

( ) This is not user-visible or docs only and no release notes are required.
( ) Release notes are required, please propose a release note for me.
( ) Release notes are required, with the following suggested text:

```markdown
* Fix some things. ({issue}`issuenumber`)
```

### CODE REVIEW

File: trino/client.py [hunk 1]
```diff
@@ -850,123 +850,195 @@ def map(self, rows):
...
+
+
+class RowValueMapper(ValueMapper[Tuple[Optional[Any], ...]]):
+    def __init__(self, mappers: List[ValueMapper[Any]]):
+        self.mappers = mappers
+
+    def map(self, values: List[Any]) -> Optional[Tuple[Optional[Any], ...]]:
+        if values is None:
+            return None
+        return tuple(self.mappers[idx].map(value) for idx, value in enumerate(values))
```
Comment (hashhar) [inline trino/client.py:982]: `idx` is no better than `i`. If you want longer name `index` is much better.

**Review by hashhar — APPROVED** [review]
LGTM % comment.

I did not comment on code that was just refactored even though they would have been majority of my comments. I'll submit separate PR with some suggestions in that area.

**Comment by hashhar** [comment]
@ebyhr PTAL. Please avoid leaving comments on code that was moved from existing code - out of scope for this PR and I'm submitting another PR on top of this with some cleanups. will request a review there.

File: tests/integration/test_types_integration.py [hunk 2]
```diff
@@ -156,6 +156,8 @@ def test_array(trino_connection):
     SqlTest(trino_connection) \
         .add_field(sql="CAST(null AS ARRAY(VARCHAR))", python=None) \
         .add_field(sql="ARRAY['a', 'b', null]", python=['a', 'b', None]) \
+        .add_field(sql="ARRAY[1.2, 2.4, null]", python=[Decimal("1.2"), Decimal("2.4"), None]) \
```
Comment (ebyhr) [inline tests/integration/test_types_integration.py:327]: "Refactor value mappers to separate classes"

"Refactoring" commit should focus on refactoring. It would be better to change the commit title.

File: tests/integration/test_types_integration.py [hunk 3]
```diff
@@ -142,6 +145,171 @@ def test_json(trino_connection):
...
+    SqlTest(trino_connection) \
+        .add_field(sql="CAST(null AS TIME)", python=None) \
+        .add_field(sql="CAST(null AS TIME(0))", python=None) \
+        .add_field(sql="CAST(null AS TIME(3))", python=None) \
+        .add_field(sql="CAST(null AS TIME(6))", python=None) \
+        .add_field(sql="CAST(null AS TIME(9))", python=None) \
+        .add_field(sql="CAST(null AS TIME(12))", python=None) \
+        .add_field(sql="CAST('01:23:45' AS TIME(0))", python=time_0) \
+        .add_field(sql="TIME '01:23:45.123'", python=time_3) \
+        .add_field(sql="CAST('01:23:45.123' AS TIME(3))", python=time_3) \
```
Comment (ebyhr) [inline tests/integration/test_types_integration.py:176]: What we want to test is type mapping between Python client and Trino server, not literal and cast in Trino server. I suppose we can remove either one. If we still want to test both, `test_date` misses the cast coverage.

**Review by ebyhr — APPROVED** [review]
Left only minor comments.

File: tests/integration/test_types_integration.py (hunk 3 above)
Comment (mdesmet) [reply tests/integration/test_types_integration.py]: It is indeed useless. I have removed the casts also we should not test precision of null values so I kept only one null test for each data type.

File: tests/integration/test_types_integration.py (hunk 2 above)
Comment (mdesmet) [reply tests/integration/test_types_integration.py:297]: Done

File: trino/client.py [hunk 4]
```diff
@@ -869,82 +869,79 @@ def map(self, value) -> Optional[float]:
...
-class TimeValueMapper(ValueMapper[time], TemporalValueMapper):
-    def __init__(self, column):
-        pattern = "%H:%M:%S"
-        ms_size, ms_to_trim = self._get_number_of_digits(column)
-        if ms_size > 0:
-            pattern += ".%f"
-        self.pattern = pattern
-        self.time_size = 9 + ms_size - ms_to_trim
+def round_millis(value: str, precision: int) -> str:
+    if len(value) > precision:
```
Comment (hashhar) [inline trino/client.py:873]: This makes assumption that input is a non-fractional number? Let's defensively verify our assumption. Nothing prevents the caller from using the function in other contexts in future.

File: trino/client.py [hunk 5]
```diff
@@ -869,82 +869,79 @@ def map(self, value) -> Optional[float]:
...
+        return str(round(Decimal(value) / Decimal(10 ** (len(value) - precision))))
```
Comment (hashhar) [inline trino/client.py:874]: It seems to perform half-odd rounding while docs say `round` does half-even. (https://docs.python.org/3/library/functions.html#round)

> if two multiples are equally close, rounding is done toward the even choice

```python
In [18]: round(Decimal('2.5'))
Out[18]: 2

In [19]: round(Decimal('3.5'))
Out[19]: 4
```

```python
In [23]: round_millis('1234585', 6)
Out[23]: '123458'

In [24]: round_millis('1234575', 6)
Out[24]: '123458'
```

Also what if the value rounds up past 1 second? Receiver needs to perform the carry.

e.g. `round_millis('9999999', 6) -> '1000000'`.

File: trino/client.py [hunk 6]
```diff
@@ -869,82 +869,79 @@ def map(self, value) -> Optional[float]:
...
-
-class TimeValueMapper(ValueMapper[time], TemporalValueMapper):
-    def __init__(self, column):
-        pattern = "%H:%M:%S"
-        ms_size, ms_to_trim = self._get_number_of_digits(column)
-        if ms_size > 0:
-            pattern += ".%f"
-        self.pattern = pattern
-        self.time_size = 9 + ms_size - ms_to_trim
+def round_millis(value: str, precision: int) -> str:
```
Comment (hashhar) [inline trino/client.py:872]: It'll be unit-testable and reusable if it returns `int` and conversion to `str` is left to caller.

File: trino/client.py [hunk 7]
```diff
@@ -869,82 +869,79 @@ def map(self, value) -> Optional[float]:
...
+        if timezone_str.startswith("-"):
+            return timezone(-timedelta(hours=int(hours), minutes=int(minutes)))
+        return timezone(timedelta(hours=int(hours), minutes=int(minutes)))
+    else:
+        return pytz.timezone(timezone_str)
+
+
+class TimeValueMapper(ValueMapper[time]):
+    def __init__(self, precision):
+        self.time_default_size = 9  # size of 'HH:MM:SS.' (the time string up to the milliseconds)
```
Comment (hashhar) [inline trino/client.py:891]: Why include the `.` in the length. I assume to make it easier to use slicing?

**Review by hashhar — APPROVED** [review]
LGTM. Timestamp changes have been extracted.
